allennlp/data/dataset_readers/srl_bert_reader.py

```diff
@@ -0,0 +1,143 @@
+                 lazy: bool = False) -> None:
+
+        if token_indexers:
+            raise ConfigurationError("The SrlBertReader has a fixed input "
+                                     "representation. Do not pass a token_indexer.")
+        super().__init__(token_indexers, domain_identifier, lazy)
+        self.bert_tokenizer = BertTokenizer.from_pretrained(bert_model_name)
+        self.lowercase_input = lowercase_input
+
+    def _tokenize_input(self, tokens: List[str]):
+        word_piece_tokens: List[str] = []
+        offsets = [0]
+        for token in tokens:
+            if self.lowercase_input:
+                token = token.lower()
+            word_pieces = self.bert_tokenizer.wordpiece_tokenizer.tokenize(token)
+            offsets.append(offsets[-1] + len(word_pieces))
+            word_piece_tokens.extend(word_pieces)
+        del offsets[0]
+
+        wordpieces = [START_TOKEN] + word_piece_tokens + [SEP_TOKEN]
+
+        offsets = [x + 1 for x in offsets]
+        return wordpieces, offsets
+
+    @staticmethod
+    def _convert_tags_to_wordpiece_tags(tags: List[str], offsets: List[int]):
+        # account for the fact the offsets are with respect to
+        # additional cls token at the start.
+        offsets = [x - 1 for x in offsets]
+        new_tags = []
+        j = 0
+        for i, offset in enumerate(offsets):
+            tag = tags[i]
+            is_o = tag == "O"
+            is_start = True
+            while j < offset:
+                if is_o:
+                    new_tags.append("O")
+
+                elif tag.startswith("I"):
+                    new_tags.append(tag)
+
+                elif is_start and tag.startswith("B"):
+                    new_tags.append(tag)
+                    is_start = False
+
+                elif tag.startswith("B"):
+                    _, label = tag.split("-", 1)
+                    new_tags.append("I-" + label)
+                j += 1
+
+        # Add O tags for cls and sep tokens.
+        return ["O"] + new_tags + ["O"]
+
+    @overrides
+    def text_to_instance(self,  # type: ignore
+                         tokens: List[Token],
+                         verb_label: List[int],
+                         tags: List[str] = None) -> Instance:
+        """
+        We take `pre-tokenized` input here, along with a verb label.  The verb label should be a
+        one-hot binary vector, the same length as the tokens, indicating the position of the verb
+        to find arguments for.
+        """
+        wordpieces, offsets = self._tokenize_input([t.text for t in tokens])
+
+        new_tags = self._convert_tags_to_wordpiece_tags(tags, offsets)
+        new_verbs = [1 if  "-V" in tag else 0 for tag in new_tags]
+        # In order to override the indexing mechanism, we need to set the `text_id`
+        # attribute directly. This causes the indexing to use this id.
+        token_field = TextField([Token(t, text_id=self.bert_tokenizer.vocab[t]) for t in wordpieces],
```

Comment:
Don't we have a bert token indexer?  You're not using that because it groups wordpieces into tokens?  What if we have a bert wordpiece indexer, that just uses this vocabulary and doesn't do any vocab counting, etc.?  Then you wouldn't have to have this hack, and you could just manually set `self._token_indexers = BertWordpieceIndexer` in the constructor.

Reply:
Actually, sorry, I think the right thing to do here, instead of having a different _indexer_, is to have a `BertWordpieceTokenizer` that does the tokenization as you have it and sets the text id for all wordpieces.  Then you just use the standard single id indexer, and everything is fine.

Reply:
This doesn't work because I need to get back the offsets of the wordpieces which doesn't fit the `Tokenizer.tokenize` api. Obviously I could re-compute these, or override the type. What do you suggest?

Reply:
The spacy tokenizer returns offsets as a field on each token. Can't you do the same?

Reply:
I could do that, but 1) It doesn't achieve the objective of being able to swap out tokenizers, because that `offset` must be set for what I am doing, so the only reason to do that would be that we are imagining this is quite general, and 2) it would have different semantics from the field `offset` in Spacy, which would be confusing. I can do it if you want, but it seems a bit "jumping through hoops" for something we don't know if people are going to use extensively (and if they did want to use it extensively we should really be considering a fuller re-write which places less importance on `TokenIndexers`).

Reply:
You have a lot more context here than I do; if what I suggested doesn't make sense, do what makes sense. I thought the offsets you needed were the same as what spacy gives (and squad models also require those offsets); I must have missed something.